src/serverModel.ts

```diff
@@ -1,47 +1,222 @@
+                event.CustomFlags[prop].forEach(function(customFlagProperty) {
+                    // TODO: Can we use our utility functions here?
+                    if (
+                        typeof customFlagProperty === 'number' ||
+                        typeof customFlagProperty === 'string' ||
+                        typeof customFlagProperty === 'boolean'
                     ) {
-                        if (
-                            typeof customFlagProperty === 'number' ||
-                            typeof customFlagProperty === 'string' ||
-                            typeof customFlagProperty === 'boolean'
-                        ) {
-                            valueArray.push(customFlagProperty.toString());
-                        }
-                    });
-                } else if (
-                    typeof event.CustomFlags[prop] === 'number' ||
-                    typeof event.CustomFlags[prop] === 'string' ||
-                    typeof event.CustomFlags[prop] === 'boolean'
-                ) {
-                    valueArray.push(event.CustomFlags[prop].toString());
-                }
+                        valueArray.push(customFlagProperty.toString());
+                    }
+                });
+            } else if (
+                // TODO: Can we use our utility functions here?
+                typeof event.CustomFlags[prop] === 'number' ||
+                typeof event.CustomFlags[prop] === 'string' ||
+                typeof event.CustomFlags[prop] === 'boolean'
+            ) {
+                valueArray.push(event.CustomFlags[prop].toString());
+            }
 
-                if (valueArray.length) {
-                    dto.flags[prop] = valueArray;
-                }
+            if (valueArray.length) {
+                dto.flags[prop] = valueArray;
             }
         }
     }
+}
+
+function convertProductToDTO(product: SDKProduct): IProductDTO {
+    return {
+        id: parseStringOrNumber(product.Sku),
+        nm: parseStringOrNumber(product.Name),
+        pr: parseNumber(product.Price),
+        qt: parseNumber(product.Quantity),
+        br: parseStringOrNumber(product.Brand),
+        va: parseStringOrNumber(product.Variant),
+        ca: parseStringOrNumber(product.Category),
+        ps: parseNumber(product.Position),
+        cc: parseStringOrNumber(product.CouponCode),
+        tpa: parseNumber(product.TotalAmount),
+        attrs: product.Attributes,
+    };
+}
+
+function convertProductListToDTO(productList: SDKProduct[]): IProductDTO[] {
+    if (!productList) {
+        return [];
+    }
+
+    return productList.map(function(product) {
+        return convertProductToDTO(product);
+    });
+}
 
-    this.appendUserInfo = function(user, event) {
+export default function ServerModel(
+    this: IServerModel,
+    mpInstance: MParticleWebSDK
+) {
+    var self = this;
+
+    // TODO: Can we refactor this to not mutate the event?
```

Comment:
yes

Reply:
Set up for future work here: https://go.mparticle.com/work/SQDSDKS-4813